Nanopore Leukemia Classification
================================

This respository includes code for building the training database, performing cross-validation, and basic performance analysis of nanopore sequencing-based transcriptomics classification of pediatric acute leukemias. This project is ongoing and the software here represents the analysis as published in *Wang et al. Transcriptional Profiling Using Low-Cost Nanopore Sequencing for Acute Leukemia Classification. (under review)*

Installation
------------

Tools to build the gene expression database and perform cross-validation are written in Python (3.6+)

Dependencies:
    numpy
    pandas
    imblearn.over\_sampling
    sklearn.ensemble
    sklearn.cross\_decomposition
    sklearn.preprocessing
    sklearn.svm
    tensorflow (Keras)

You will also need to download the ENSEMBL cDNA reference and associated metadata:

Ex. for release 105:
    wget http://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz

An appropriate ENSEMBL metadata file can be generated by (https://m.ensembl.org/biomart/martview/)[biomart] data dump with the following fields in CSV format:

    Gene stable ID
    Gene stable ID version
    Transcript stable ID
    Transcript stable ID version
    Gene name
    Gene Synonym
    Chromosome/scaffold name
    Gene start (bp)
    Gene end (bp)
    Gene description
    Transcript start (bp)
    Transcript end (bp)
    Transcript name
    HGNC symbol

Usage
-----

Generate transcript counts from nanopore sequence using (https://github.com/txje/minnow)[minnow]:
    Usage: minnow [options]
    Options:
      -q: FASTA/Q[.gz] file with reads
      -r: Reference FASTA/Q[.gz]
      -t: Threads (default: 1)
      -c: 'careful': more accurate but slower
      -v, --verbose: verbose
      -h, --help: show this
      --version: show version information

For example:
    minnow -q nanopore_reads.fastq.gz -r Homo_sapiens.GRCh38.cdna.all.fa.gz > nanopore_transcripts.cts

Generate an appropriate metadata file (metadata.tsv) in the following format:
    run_name	sample_name	file	lineage	type_1	type_2	type_3	subtype_note	attr_diagnosis	attr_subtype_biomarkers	subject_name	SJ_ID
    np1	sample_A	nanopore_transcripts.cts	B-ALL	KMT2Ar	NA	NA		Acute B-cell Leukemia	KMT2A-MLLT10	subject_A	external_id_A
    ...

Build gene expression matrix for training and cross-validation:
    python build_training_db2.py <metadata.tsv> <gene expression matrix (numpy)> <gene list (text)> <sample names (text)> <ENSEMBL gene data>

Example:
    python build_training_db2.py data/metadata.tsv data/gexp_matrix.npy data/gene_list.txt data/sample_names.txt data/ensembl_105_grch38.tsv

Run cross-validation:
    python classify.py <data_dir> <data_domain> <smote>
      data_dir: Path to directory with data. Directory must contain gexp_matrix.npy, metadata.tsv, sample_names.txt, and gene_list.txt
      data_domain: Set of data to use. Options are major, AML, or B-ALL.
      smote: Whether or not to use SMOTE to balance the dataset. Options are True or False (or 0/1). SMOTE is a data augmentation algorithm that generates artificial samples of all minority classes.
